app/models/exercise.rb

```diff
@@ -300,35 +366,108 @@ def determine_file_role_from_proforma_file(task_node, file_node)
     return 'regular_file'
   end
 
-  def from_proforma_xml(xml_string)
-    # how to extract the proforma functionality into a different module in rails?
-    xml = Nokogiri::XML(xml_string)
-    xml.collect_namespaces
-    task_node = xml.xpath('/root/p:task')
-    description = task_node.xpath('p:description/text()')[0].content
-    self.attributes = {
-      title: task_node.xpath('p:meta-data/p:title/text()')[0].content,
-      description: description,
-      instructions: description
+  def build_proforma_xml_for_exercise_file(xml, exercise_file)
+    if exercise_file.role == 'main_file'
+      proforma_file_class = 'template'
+      comment = 'main'
+    else
+      proforma_file_class = 'internal'
+      comment = ''
+    end
+
+    xml['p'].file(exercise_file.content,
+                  'filename' => exercise_file.full_file_name,
+                  'id' => exercise_file.id,
+                  'class' => proforma_file_class,
+                  'comment' => comment
+    )
+  end
+
+  def build_proforma_xml_for_test(xml, test, index)
+    proforma = xml['p']
+    proforma.test('id' => 't' + index.to_s) {
+      proforma.title('')
+      proforma.send('test-type', 'unittest')
+      proforma.send('test-configuration') {
+        proforma.filerefs {
+          proforma.fileref('refid' => test.id.to_s)
+        }
+        xml['u'].unittest('framework' => 'JUnit 4', 'version' => '')
+        xml['c'].send('feedback-message', test.feedback_message)
+      }
     }
-    task_node.xpath('p:files/p:file').all? { |file|
-      file_name_split = file.xpath('@filename').first.value.split('.')
-      file_class = file.xpath('@class').first.value
-      role = determine_file_role_from_proforma_file(task_node, file)
-      feedback_message_nodes = task_node.xpath("p:tests/p:test/p:test-configuration/c:feedback-message/text()")
-      files.build({
-        name: file_name_split.first,
-        content: file.xpath('text()').first.content,
-        read_only: false,
-        hidden: file_class == 'internal',
-        role: role,
-        feedback_message: (role == 'teacher_defined_test') ? feedback_message_nodes.first.content : nil,
-        file_type: FileType.where(
-          file_extension: ".#{file_name_split.second}"
-        ).take
-      })
+  end
+
+  def build_proforma_xml_for_model_solution(xml, model_solution_file, index)
+    proforma = xml['p']
+    proforma.send('model-solution', 'id' => 'm' + index.to_s) {
+      proforma.filerefs {
+        proforma.fileref('refid' => model_solution_file.id.to_s)
+      }
     }
-    self.execution_environment_id = 1
+  end
+
+  def to_proforma_xml
+    builder = Nokogiri::XML::Builder.new do |xml|
+      proforma = xml['p']
+      proforma.task('xmlns:p' => 'urn:proforma:task:v1.1', 'lang' => 'en', 'uuid' => SecureRandom.uuid,
```

Comment:
change the default language here to German, please. Most of our exercises are German, currently, so it will be less work to adjust the others.